library/pem.c

```diff
@@ -239,16 +274,17 @@ int mbedtls_pem_read_buffer( mbedtls_pem_context *ctx, const char *header, const
     ((void) pwdlen);
 #endif /* MBEDTLS_MD5_C && MBEDTLS_CIPHER_MODE_CBC &&
           ( MBEDTLS_AES_C || MBEDTLS_DES_C ) */
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
 
     if( ctx == NULL )
         return( MBEDTLS_ERR_PEM_BAD_INPUT_DATA );
 
-    s1 = (unsigned char *) strstr( (const char *) data, header );
+    s1 = (unsigned char *) mbedtls_pem_strnstr( (const char *) data, header, datalen );
 
     if( s1 == NULL )
         return( MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT );
 
-    s2 = (unsigned char *) strstr( (const char *) data, footer );
+    s2 = (unsigned char *) mbedtls_pem_strnstr( (const char *) data, footer, datalen );
 
```

Comment:
Now the code that skips whitespace after the header and footer can overflow the input buffer. For the header, it can only happen in pathological cases where the footer consists only of whitespace, but that's not a documented exclusion. For the footer, it's quite likely, and in fact the CI found it with `tests/data_files/cli-rsa.key`.

Please fix this and add unit tests: `pk_parse_key` on `tests/data_files/cli-rsa.key`, and some more general test cases on `mbedtls_pem_read_buf_with_len` directly.

(That's what's causing most of the CI failures.)